spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkBatchQueryScan.java

```diff
@@ -74,60 +101,158 @@
-        } else {
-          scan = scan.appendsAfter(startSnapshotId);
-        }
+    for (Integer specId : specIds()) {
+      PartitionSpec spec = table().specs().get(specId);
+      for (PartitionField field : spec.fields()) {
+        partitionFieldSourceIds.add(field.sourceId());
       }
+    }
 
-      if (splitSize != null) {
-        scan = scan.option(TableProperties.SPLIT_SIZE, splitSize.toString());
-      }
+    Map<Integer, String> quotedNameById = TypeUtil.indexQuotedNameById(
+        expectedSchema().asStruct(),
+        name -> String.format("`%s`", name.replace("`", "``")));
 
-      if (splitLookback != null) {
-        scan = scan.option(TableProperties.SPLIT_LOOKBACK, splitLookback.toString());
-      }
+    return partitionFieldSourceIds.stream()
+        .filter(fieldId -> expectedSchema().findField(fieldId) != null)
+        .map(fieldId -> Spark3Util.toNamedReference(quotedNameById.get(fieldId)))
+        .toArray(NamedReference[]::new);
+  }
 
-      if (splitOpenFileCost != null) {
-        scan = scan.option(TableProperties.SPLIT_OPEN_FILE_COST, splitOpenFileCost.toString());
+  @Override
+  public void filter(Filter[] filters) {
+    Expression runtimeFilterExpr = convertRuntimeFilters(filters);
+
+    if (runtimeFilterExpr != Expressions.alwaysTrue()) {
+      Map<Integer, Evaluator> evaluatorsBySpecId = Maps.newHashMap();
+
+      for (Integer specId : specIds()) {
+        PartitionSpec spec = table().specs().get(specId);
+        Expression inclusiveExpr = Projections.inclusive(spec).project(runtimeFilterExpr);
+        Evaluator inclusive = new Evaluator(spec.partitionType(), inclusiveExpr);
+        evaluatorsBySpecId.put(specId, inclusive);
       }
 
-      for (Expression filter : filterExpressions()) {
-        scan = scan.filter(filter);
+      LOG.info("Trying to filter {} files using runtime filter {}", files().size(), runtimeFilterExpr);
+
+      List<FileScanTask> filteredFiles = files().stream()
+          .filter(file -> {
+            Evaluator evaluator = evaluatorsBySpecId.get(file.spec().specId());
+            return evaluator.eval(file.file().partition());
+          })
+          .collect(Collectors.toList());
+
+      LOG.info("{}/{} files matched runtime filter {}", filteredFiles.size(), files().size(), runtimeFilterExpr);
+
+      // don't invalidate tasks if the runtime filter had no effect to avoid planning splits again
+      if (filteredFiles.size() < files().size()) {
+        this.specIds = null;
+        this.files = filteredFiles;
+        this.tasks = null;
       }
 
-      try (CloseableIterable<CombinedScanTask> tasksIterable = scan.planTasks()) {
-        this.tasks = Lists.newArrayList(tasksIterable);
-      }  catch (IOException e) {
-        throw new RuntimeIOException(e, "Failed to close table scan: %s", scan);
+      runtimeFilterExpressions.add(runtimeFilterExpr);
+    }
+  }
+
+  private Expression convertRuntimeFilters(Filter[] filters) {
```

Comment:
Side question: do we need to do any analysis of the incoming filters? For example, if there is one ID field will the filter be a series of ANDed equality predicates that we can convert to an IN?

We also don't currently support an IN predicate of StructLike values, but we could implement that if it is something we expect. Maybe we should document what the expected filters look like in a comment?

Reply:
I'll add a comment. At this moment, we can only get an array of `IN` filters. That reminds me to add a test case for filtering on multiple partition columns.

Reply:
Added a comment.